analysis_schema/data_classes.py

```diff
@@ -1,34 +1,36 @@
 from pathlib import Path
 from typing import List, Optional, Tuple, Union
 
-import yt
 from pydantic import BaseModel, Field
 
-from ._data_store import _instantiated_datasets
+from ._data_store import dataset_fixture
 from .base_model import ytBaseModel, ytDataObjectAbstract, ytParameter
 
 
 class Dataset(ytBaseModel):
     """
     The dataset to load. Filename (fn) must be a string.
 
-    Required fields: Filename
+    Required fields: Filename, DatasetName
     """
 
+    DatasetName: Optional[str]
     fn: Path = Field(
         alias="FileName",
         description="A string containing the (path to the file and the) file name",
     )
-    DatasetName: Optional[str]
     comments: Optional[str]
+    instantiate: bool = True
     _yt_operation: str = "load"
 
     def _run(self):
```

Comment:
Ok, I think I understand but I have a question. I've added a check for `DatasetName is None` and removed the `instantiate` attribute. Here's what it looks like: 

```
def _run(self):
        if self.DatasetName is not None:
            if self.DatasetName in [dataset_fixture._instantiated_datasets.keys()]:
                return dataset_fixture._instantiated_datasets[self.DatasetName]
            else:
                dataset_fixture.add_to_alldata(self.fn, self.DatasetName)
                ds = dataset_fixture._instantiate_data(self.DatasetName)
                return ds
        else:
            raise AttributeError("Missing a dataset!")
```

And I understand what you mean by setting `self.DatasetName = self.fn` to make sure a duplicate file isn't added to the dictionary, but the `DatasetName` is meant to be the key and `fn` to be the value. The key shouldn't be an integer (let me double check this though, because at one point I was doing that as  quick fix). Are you suggesting the filename be the key and value in the dict?

I've can also make `DatasetName` required in the schema so won't be validated without that attribute. 

Thoughts?

Reply:
> Are you suggesting the filename be the key and value in the dict? 

Yup! but only if the user has not set something for `DatasetName`. 

And the new code looks good! If you want, you could make it more compact by instead first checking if `DatasetName` is missing, then adding it right away so that it's guaranteed to exist:

```python
        if self.DatasetName is not None:
            if self.DatasetName not in dataset_fixture._instantiated_datasets:  
                dataset_fixture.add_to_alldata(self.fn, self.DatasetName)
            return dataset_fixture._instantiated_datasets[self.DatasetName]
        else:
             raise AttributeError("Missing a dataset!")
```

Reply:
Oh, I think I misunderstood your question -- I meant that the the key should be `DatasetName` and in the case where a `DatasetName` is not provided, set `DatasetName = the file name`, in which case the filename would be both a key and a value.

Reply:
There's this bit of code in the `_data_store.DatasetFixture` class:

```python
    def add_to_alldata(self, fn: str, dataset_name: str):
        """
        A function to track all dataset.
        Stores dataset name, or if no name is provided,
        adds a number as the name.
        """
        self.fn = fn
        if dataset_name is not None:
            self.dataset_name = dataset_name
        else:
            self.dataset_name = len(self.all_data.values())
```
and I'm saying that that if statement could be:

```python
        if dataset_name is not None:
            self.dataset_name = dataset_name
        else:
            self.dataset_name = fn
```

Reply:
or maybe that should happen earlier....